orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py

```diff
@@ -0,0 +1,128 @@
+
+# opset 10
+from ort_supplement.onnx_transforms.model_transform import add_name, fix_transpose, add_expand_shape, process_concat, process_dropout, handle_expand_input_is_not_constant_case, fix_dim, fix_expand
+
+from ort_supplement.onnx_transforms.layer_norm_transform import layer_norm_transform
+
+def postprocess_model(model):
+    add_name(model)
+
+    # not in TJ branch process_concat(model)
+
+    # opset 10:
+    handle_expand_input_is_not_constant_case(model)
+    fix_expand(model)
+    fix_dim(model)
+    process_dropout(model)
+    # --- 
+
+    # not in TJ branch fix_transpose(model)
+    add_expand_shape(model)
+    layer_norm_transform(model)
+
+def create_ort_trainer(args, device, model):
+    # set GPU memory limitation
+    from onnxruntime.capi._pybind_state import set_cuda_mem_limit
+    ort_cuda_mem_limit_in_gbs = 16
+    set_cuda_mem_limit(int(ort_cuda_mem_limit_in_gbs * 1024 * 1024 *1024))
+
+    # BertLAMB default initial settings: b1=0.9, b2=0.999, e=1e-6
+    def map_optimizer_attributes(name):
+        no_decay_keys = ["bias", "gamma", "beta", "LayerNorm"]
+        no_decay = False
+        for no_decay_key in no_decay_keys:
+            if no_decay_key in name:
+                no_decay = True
+                break
+        if no_decay:
+            return {"alpha": 0.9, "beta": 0.999, "lambda": 0.0, "epsilon": 1e-6}
+        else:
+            return {"alpha": 0.9, "beta": 0.999, "lambda": 0.01, "epsilon": 1e-6}
+
+    # we request ORTTrainer to create a LambOptimizer with given optimizer_attributes. 
+    # train_step does forward, backward, and optimize step.
+    model = ORTTrainer(model, None, bert_model_description(args), "LambOptimizer", 
+        map_optimizer_attributes,
+        IODescription('Learning_Rate', [1,], torch.float32),
+        device, postprocess_model=postprocess_model, 
+        gradient_accumulation_steps=args.gradient_accumulation_steps,
+        world_rank=args.world_rank, world_size=args.world_size,
+        use_mixed_precision = True if args.fp16 else False,
+        allreduce_post_accumulation = True if args.allreduce_post_accumulation else False,
+        partition_optimizer=True,
+        _opset_version = 10)
+
+    return model
+
+from ort_supplement.lr_schedules import SCHEDULES
+def get_lr(args, training_steps, schedule='warmup_poly'):
+    if args.max_steps == -1:
+        return args.learning_rate
+
+    schedule_fct = SCHEDULES[schedule]
+    return args.learning_rate * schedule_fct(training_steps / args.max_steps, args.warmup_proportion)
+
+def run_ort_training_step(args, global_step, training_steps, model, batch):
+    input_ids, segment_ids, input_mask, masked_lm_labels, next_sentence_labels = batch
+
+    if args.fp16:
+        loss_scaler = LossScaler(model.loss_scale_input_name, True, up_scale_window=2000)
```

Comment:
There's another issue here. This means every step we will create a new lossScalar with a large magnitude, which means the gradients will keep overflowing and weights will never be updated. I will suggest moving this to create_ort_trainer and add it as an attr to args, i.e.:
```
if args.fp16:
    setattr(args, 'ort_loss_scale', LossScaler(model.loss_scale_input_name, True, up_scale_window=2000))
return model
```
Then in run_ort_training_step you can just do
```
loss_scale = torch.tensor([args.ort_loss_scale.loss_scale_])
```
and
```
if args.fp16:
    args.ort_loss_scale.update_loss_scale(all_finite.item())
```

Reply:
I made these fixes in PR at https://github.com/microsoft/onnxruntime-training-examples/pull/2